source/contributing/implementation/datamanipulator.rst

```diff
@@ -0,0 +1,272 @@
+custom entity/itemstack/tileentity/blockstate types. The higher the priority return, the `earlier` the processor is used
+in what's called the ``DataProcessorDelegate``. The delegate is a collection of registered ``DataProcessor`` of a
+particular ``DataManipulator`` such that all registered ``DataProcessor`` can be used.
+
+supports(DataHolder)
+--------------------
+
+Next up, we have ``supports(DataHolder)``. This is really simple, if your ``DataManipulator`` is supposed to be
+"supported" by the provided ``DataHolder`` (which can be an ItemStack, Entity, TileEntity, etc.) you can define the
+"support" in this method. So, if I have a ``VelocityDataProcessor``, my ``support(DataHolder)`` would look something like so:
+
+.. code-block:: java
+
+ @Override
+ public boolean supports(DataHolder dataHolder) {
+  return dataHolder instanceof Entity;
+ }
+
+from(DataHolder)
+----------------
+
+createFrom(DataHolder)
+----------------------
+
+.. warning::
+
+ Not sure the code is right. Please doublecheck!
+ Code taken out of PR#139 for now.
+
+.. code-block:: java
+
+ public Optional<FlyingData> createFrom(DataHolder dataHolder) {
+        if (supports(dataHolder)) {
+            boolean isFlying = ((net.minecraft.entity.Entity) dataHolder).isAirBorne;
+            if (dataHolder instanceof EntityPlayer) {
+                isFlying = ((EntityPlayer) dataHolder).capabilities.allowFlying;
+            }
+            return Optional.<FlyingData>of(new SpongeFlyingData(isFlying));
+        }
+        return Optional.absent();
+    }
+
+fill(DataHolder)
+----------------
+
+.. warning::
+
+  Not sure the code is right, especially the ``checkNotNull`` part. Please doublecheck!
+  Code taken out of PR#139 for now.
+
+.. code-block:: java
+
+    public Optional<FlyingData> fill(DataHolder dataHolder, FlyingData manipulator, MergeFunction overlap) {
+        if (supports(dataHolder)) {
+            final FlyingData merged = checkNotNull(overlap.merge(checkNotNull(manipulator).copy(), from(dataHolder).get()), "MergeFunction cannot be null!");
+            manipulator.set(Keys.IS_FLYING, merged.flying().get());
+            return Optional.of(manipulator);
+        }
+        return Optional.absent();
+    }
+
+fill(DataContainer)
+-------------------
+
+.. warning::
+
+ Not sure the code is right. Please doublecheck!
+ Code taken out of PR#139 for now.
+
+.. code-block:: java
+
+ public Optional<FlyingData> fill(DataContainer container, FlyingData flyingData) {
+        flyingData.set(Keys.IS_FLYING, getData(container, Keys.IS_FLYING));
+        return Optional.of(flyingData);
+ }
+
+set(DataHolder)
+---------------
+
+with(Key)
+---------
+
+remove(DataHolder)
+------------------
+
+.. warning::
+
+ Not sure the code is right. Please doublecheck!
+ Code taken out of PR#139 for now.
+
+.. code-block:: java
+
+ public DataTransactionResult remove(DataHolder dataHolder) {
```

Comment:
There should probably be more explaining on when it's appropriate to not implement remove.